prusti-viper/src/encoder/snapshot_encoder.rs

```diff
@@ -177,23 +180,27 @@ impl<'p, 'v, 'r: 'v, 'a: 'r, 'tcx: 'a> SnapshotEncoder<'p, 'v, 'tcx> {
             | ty::TyKind::Uint(_)
             | ty::TyKind::Char
             | ty::TyKind::Bool => {
-                self.encode_snap_primitive(
+                self.encode_primitive(
                     self.encoder.encode_value_field(self.ty)
                 )
             }
             ty::TyKind::Param(_) => {
-                self.encode_snap_generic()?
-            }
-            ty::TyKind::Adt(adt_def, _) if !adt_def.is_box() => {
-                if adt_def.variants.len() != 1 {
-                    warn!("Generating equality tests for enums is not supported");
-                }
-                self.encode_snap_struct()?
+                self.encode_generic()?
             }
             ty::TyKind::Tuple(_) => {
-                self.encode_snap_struct()?
+                self.encode_tuple()?
+            }
+            ty::TyKind::Adt(adt_def, subst) if !adt_def.is_box() => {
```

Comment:
Wouldn't a `box` just be a recursive call to the encoder, or is there anything special about it? I assume something like `Box<dyn ...>` should fail in the recursive call because of the `dyn`, not because of the box.

Reply:
[Box](https://doc.rust-lang.org/src/alloc/boxed.rs.html#160) is a wrapper around a [raw pointer](https://doc.rust-lang.org/1.21.0/src/core/ptr.rs.html#1082-1090). Our encoding is a bit special (see below) and I think the compiler special-cases it too.

https://github.com/viperproject/prusti-dev/blob/ee5def9423f99af974300d8d7d1a43377354e8cf/prusti-viper/src/encoder/type_encoder.rs#L360-L368